### PR TITLE
VIMC-3959: Port orderly runner to orderly.server

### PR DESCRIPTION
**Pull request by r-ash**
This definitely requires some discussion.

In general I have tried to change as little of the implementation here as possible. And this is mostly just straight copies from orderly. This has resulted in quite a lot of duplication some of which is probably fine but others I'm not sure if we want to try and avoid the duplication and decide on some approach for sharing code. I'll try and highlight those bits in particular.

Some remaining code factor issues, some are TODO's others are places were we are using unexported functions from orderly.

### CODE REVIEW

File: R/util.R [hunk 1]
```diff
@@ -33,3 +33,121 @@ vcapply <- function(X, FUN, ...) { # nolint
...
+    filename <- paste0(basename(file), suffix)
+  }
+  file.path(root, "backup", "db", filename, fsep = "/")
+}
+
+path_draft <- function(root) {
+  file.path(root, "draft")
+}
+
+path_archive <- function(root, name = NULL) {
```
[the remark this answers is not in the document]
Comment (r-ash) [reply R/util.R:127]: Sounds good, I've moved `path_orderly_run_rds` and `path_db_backup` to use `orderly:::path_orderly_run_rds` in the helper function like we do for other imports from orderly which are only used in tests. I will leave the others as dupes for now, the `path_runner_*` which can be removed from `orderly` soon which sounds good.

File: R/runner.R [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+      if (!self$has_git) {
+        message("Not enabling git features as this is not version controlled")
+      }
+
+      self$allow_ref <- runner_allow_ref(self$has_git, allow_ref, self$config)
+      if (self$has_git && !self$allow_ref) {
+        message("Disallowing reference switching in runner")
+      }
+
+      do_backup <- protect(function() orderly:::orderly_backup(self$config))
```
[the remark this answers is not in the document]
Comment (r-ash) [reply R/runner.R:81]: Yep, created a ticket https://mrc-ide.myjetbrains.com/youtrack/issue/VIMC-4046

File: R/util.R [hunk 3]
```diff
@@ -33,3 +33,108 @@ vcapply <- function(X, FUN, ...) { # nolint
...
+  if (is.null(name)) {
+    file.path(root, "archive")
+  } else {
+    file.path(root, "archive", name)
```
Comment (richfitz) [inline R/util.R:122]: This path is never used

**Comment by r-ash** [comment]
Trying to mock the processx kill call proving pretty tricky as you suspected so I'm just stopping on this. This is good for another look I think. Only code factor failing on TODO lints, we're ignoring the TODO lintr on master so I'm not sure why they are still being reported. Any ideas?

**Comment by richfitz** [comment]
The `TODO` lintr is a bit of a mystery but I've seen this before. At some point let's look into it. Ultimately the point of the tool is to point things out not get in our way, and I think we're ok with those TODOs

**Pull request merged**